modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -0,0 +1,1788 @@
+
+                ctx.resource().inject(cp);
+
+                return cp;
+            }
+            catch (IgniteCheckedException e) {
+                U.error(log, "Failed to copy service (will reuse same instance): " + srvc.getClass(), e);
+
+                return srvc;
+            }
+        }
+    }
+
+    /**
+     * @param ctxs Contexts to cancel.
+     * @param cancelCnt Number of contexts to cancel.
+     */
+    private void cancel(Iterable<ServiceContextImpl> ctxs, int cancelCnt) {
+        for (Iterator<ServiceContextImpl> it = ctxs.iterator(); it.hasNext(); ) {
+            cancel(it.next());
+
+            it.remove();
+
+            if (--cancelCnt == 0)
+                break;
+        }
+    }
+
+    /**
+     * Perform cancelation on given service context.
+     *
+     * @param ctx Service context.
+     */
+    private void cancel(ServiceContextImpl ctx) {
+        // Flip cancelled flag.
+        ctx.setCancelled(true);
+
+        // Notify service about cancellation.
+        Service srvc = ctx.service();
+
+        if (srvc != null) {
+            try {
+                srvc.cancel(ctx);
+            }
+            catch (Throwable e) {
+                U.error(log, "Failed to cancel service (ignoring) [name=" + ctx.name() +
+                    ", execId=" + ctx.executionId() + ']', e);
+
+                if (e instanceof Error)
+                    throw e;
+            }
+            finally {
+                try {
+                    this.ctx.resource().cleanup(srvc);
+                }
+                catch (IgniteCheckedException e) {
+                    U.error(log, "Failed to clean up service (will ignore): " + ctx.name(), e);
+                }
+            }
+        }
+
+        // Close out executor thread for the service.
+        // This will cause the thread to be interrupted.
+        ctx.executor().shutdownNow();
+
+        if (log.isInfoEnabled()) {
+            log.info("Cancelled service instance [name=" + ctx.name() + ", execId=" +
+                ctx.executionId() + ']');
+        }
+    }
+
+    /**
+     * Undeployes service with given id.
+     * <p/>
+     * Invokes from services deployment worker.
+     *
+     * @param srvcId Service id.
+     */
+    protected void undeploy(@NotNull IgniteUuid srvcId) {
+        Collection<ServiceContextImpl> ctxs = locServices.remove(srvcId);
+
+        if (ctxs != null) {
+            synchronized (ctxs) {
+                cancel(ctxs, ctxs.size());
+            }
+        }
+    }
+
+    /**
+     * @param deploy {@code true} if complete deployment requests, otherwise complete undeployment request will be
+     * completed.
+     * @param reqSrvcId Request's service id.
+     * @param err Error to complete with. If {@code null} a future will be completed successfully.
+     */
+    protected void completeInitiatingFuture(boolean deploy, IgniteUuid reqSrvcId, Throwable err) {
```

Comment:
Why `protected`?
I actually see a lot of `protected` in the patch, and one usually needs that only when 3rd party code (or test) is supposed to extend one's classes. Could you please review the `protected` usages and replace them with `private`/package-private where possible?

Reply:
Fixed.